### PR TITLE
fix(a380x): Port Over the New A32NX GPWS to the A380X

### LINKED ISSUE #8802: [A380X] Update GPWS with recent changes from A32NX GPWS
### Aircraft Version

Development

### Description

https://github.com/flybywiresim/aircraft/pull/8136

Port to A380X's GPWS. 

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### PR DESCRIPTION
**Pull request by CaptCrystal**
Fixes #8802 

## Summary of Changes

Removed the option for a landing flaps 3 configuration.
Re-did the a380x's GPWS mode 4 logic and replaced it with new mode 4 logic. 
Added in various different fixes to the GPWS logic from the new GPWS update.

## References
image
image

Discord username (if different from GitHub):
Capt. Crystal

## Testing instructions
1. Load into an approach in the A380, any airport should work.
2. Confirm that the necessary GPWS callouts are announced at the correct altitude AGL. I recommend using flight recorder off of Flightsim.to so you can pause when you hear the callout to verify that it was made at the correct altitude AGL.
3. End flight
4. Load into the A380 in a mountainous area. 
5. Confirm that the correct GPWS terrain callouts are made. The "correct" callouts are shown in the Reference section.

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

**Comment by tracernz** [comment]
We already have a port of the GPWS code https://github.com/flybywiresim/aircraft/blob/master/fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts. It just doesn't have the latest enhancements.

**Comment by CaptCrystal** [comment]
Okay, I saw an issue that requested that someone ports over the recent updates.

**Comment by tracernz** [comment]
The GPWS was ported across in https://github.com/flybywiresim/aircraft/pull/8439. Since then https://github.com/flybywiresim/aircraft/pull/8136 was merged to the A32NX. I believe the idea is to port those changes to the A380X.

**Comment by CaptCrystal** [comment]
Okay. Are those changes in #8694 as well as others already implemented?

**Comment by tracernz** [comment]
Yes there are, so we can’t just take the A32NX GPWS directly across again.

**Comment by CaptCrystal** [comment]
Okay, may we take this into discord so as to not clutter up this PR?

**Review by flogross89 — DISMISSED** [review]

**Comment by BravoMike99** [comment]
I don't think you should be deleting the GPWS FLAP 3 checks. While the A380 does not have a dedicated button (unlike the A320), the FMS does send the selected flap setting on the perf approach page to the FMS, in other words, if you select flaps full on the fms and land with flaps three, it would scream "TOO LOW FLAPS". You can use the A32NX_SPEEDS_LANDING_CONF3 simvar to know if flaps 3 has been selected on the FMS instead of the old A32NX_GPWS_FLAPS3 for example.

**Comment by flogross89** [comment]
> I don't think you should be deleting the GPWS FLAP 3 checks. While the A380 does not have a dedicated button (unlike the A320), the FMS does send the selected flap setting on the perf approach page to the FMS, in other words, if you select flaps full on the fms and land with flaps three, it would scream "TOO LOW FLAPS". You can use the A32NX_SPEEDS_LANDING_CONF3 simvar to know if flaps 3 has been selected on the FMS instead of the old A32NX_GPWS_FLAPS3 for example.

Good spot, thanks. Yes, please let's keep in the Flap 3 checks and implement it per @BravoMike99 's suggestion

**Comment by CaptCrystal** [comment]
> > I don't think you should be deleting the GPWS FLAP 3 checks. While the A380 does not have a dedicated button (unlike the A320), the FMS does send the selected flap setting on the perf approach page to the FMS, in other words, if you select flaps full on the fms and land with flaps three, it would scream "TOO LOW FLAPS". You can use the A32NX_SPEEDS_LANDING_CONF3 simvar to know if flaps 3 has been selected on the FMS instead of the old A32NX_GPWS_FLAPS3 for example.
> 
> Good spot, thanks. Yes, please let's keep in the Flap 3 checks and implement it per @BravoMike99 's suggestion

Okay, @tracernz, may I take the current A32NX GPWS code, port it over, and then port the code that you wrote over to the new copy?

**Comment by tracernz** [comment]
It would be best to just undo the flap 3 changes so we keep everything else that's been done.

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts [hunk 1]
```diff
@@ -181,33 +181,52 @@ export class LegacyGpws {
...
         this.differentiateRadioalt(radioAltValid ? radioAlt.value : NaN, deltaTime);
 
         const mda = SimVar.GetSimVarValue('L:AIRLINER_MINIMUM_DESCENT_ALTITUDE', 'feet');
         const dh = SimVar.GetSimVarValue('L:AIRLINER_DECISION_HEIGHT', 'feet');
         const phase = SimVar.GetSimVarValue('L:A32NX_FMGC_FLIGHT_PHASE', 'Enum');
 
+      this.update_maxRA(radioAlt, isOnGround);
+
+        const isOverflightDetected = this.updateOverflightState(deltaTime);
+        this.updateMode4aInhibited(isGearDownLocked, areFlapsInLandingConfig);
```
Comment (tracernz) [inline fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:218]: `areFlapsInLandingConfig` does not exist here.

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts [hunk 2]
```diff
@@ -181,33 +181,52 @@ export class LegacyGpws {
...
+
         if (
             radioAltValid && radioAlt.value >= 10 && radioAlt.value <= 2450
             && !SimVar.GetSimVarValue('L:A32NX_GPWS_SYS_OFF', 'Bool')
         ) { // Activate between 10 - 2450 radio alt unless SYS is off
-            const FlapPushButton = SimVar.GetSimVarValue('L:A32NX_GPWS_FLAPS3', 'Bool');
             const FlapPosition = SimVar.GetSimVarValue('L:A32NX_FLAPS_HANDLE_INDEX', 'Number');
             const FlapsInLandingConfig = FlapPushButton ? (FlapPosition === 3) : (FlapPosition === 4);
```
Comment (tracernz) [inline fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:225]: But does here (with a different name). If you hook the "FlapPushButton" up to whatever data we get that from on the A380 and make it available, then you've got flaps 3 sorted.

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts [hunk 3]
```diff
@@ -181,33 +181,52 @@ export class LegacyGpws {
...
+        const isTcfOperational = this.isTerrainClearanceFloorOperational(isTerrModeOff, radioAlt, isNavAccuracyHigh);
+        const isTafOperational = this.isTerrainAwarenessOperational(isTerrModeOff);
         this.UpdateAltState(radioAltValid ? radioAlt.value : NaN);
         this.differentiateRadioalt(radioAltValid ? radioAlt.value : NaN, deltaTime);
 
         const mda = SimVar.GetSimVarValue('L:AIRLINER_MINIMUM_DESCENT_ALTITUDE', 'feet');
         const dh = SimVar.GetSimVarValue('L:AIRLINER_DECISION_HEIGHT', 'feet');
         const phase = SimVar.GetSimVarValue('L:A32NX_FMGC_FLIGHT_PHASE', 'Enum');
 
+      this.update_maxRA(radioAlt, isOnGround);
```
Comment (tracernz) [inline fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:215]: Some kind of indentation issue with this line.

File: fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts (hunk 2 above)
Comment (CaptCrystal) [reply fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:225]: I connected the areFlapsInLandingConfig constant with the FlapsInLandingConfig constant.

Comment (CaptCrystal) [reply fbw-a380x/src/systems/systems-host/systems/LegacyGpws.ts:225]: Just now.

**Comment by CaptCrystal** [comment]
Did not intend for this PR to be closed. It will be re-opened after I add commits to my fork.

**Comment by CaptCrystal** [comment]
Because of pull request #8875, It'd take some more time for me to do this, as I need to adapt what I do have to fit this pull request.

**Comment by Maximilian-Reuter** [comment]
> Because of pull request #8875, It'd take some more time for me to do this, as I need to adapt what I do have to fit this pull request.

As you still intend to do further work on this, wouldn't it be better for visibility if you simply converted this PR to draft ?
That way everyone can see that you work on this but know that it isn't ready yet.

**Comment by CaptCrystal** [comment]
> > Because of pull request #8875, It'd take some more time for me to do this, as I need to adapt what I do have to fit this pull request.
> 
> As you still intend to do further work on this, wouldn't it be better for visibility if you simply converted this PR to draft ? That way everyone can see that you work on this but know that it isn't ready yet.

Will do

**Pull request closed without merging**